File: android/src/main/java/org/maplibre/reactnative/components/annotation/MLRNMarkerView.java
package org.maplibre.reactnative.components.annotation;

import android.content.Context;
import android.graphics.PointF;
import android.view.View;

import androidx.annotation.NonNull;

import org.maplibre.geojson.Point;
import org.maplibre.android.maps.MapLibreMap;
import org.maplibre.android.maps.OnMapReadyCallback;
import org.maplibre.reactnative.components.AbstractMapFeature;
import org.maplibre.reactnative.components.mapview.MLRNMapView;
import org.maplibre.reactnative.utils.GeoJSONUtils;

public class MLRNMarkerView extends AbstractMapFeature implements MarkerView.OnPositionUpdateListener, View.OnLayoutChangeListener {
    private MLRNMarkerViewManager mManager;
    private MLRNMapView mMapView;

    private View mChildView;

    private MarkerViewManager mMarkerViewManager;

    private MarkerView mMarkerView;
    private Point mCoordinate;
    private Float[] mAnchor;


    public MLRNMarkerView(Context context, MLRNMarkerViewManager manager) {
        super(context);
        mManager = manager;
    }

    @Override
    public void addView(View childView, int childPosition) {
        mChildView = childView;
    }

    public void setCoordinate(Point point) {
        mCoordinate = point;

        if (mMarkerView != null) {
            mMarkerView.setLatLng(GeoJSONUtils.toLatLng(point));
        }
    }

    public void setAnchor(float x, float y) {
        mAnchor = new Float[]{x, y};
        refresh();
    }

    public void refresh() {
        // this will cause position to be recalculated
        if (mMarkerView != null) {
            mMarkerView.setLatLng(GeoJSONUtils.toLatLng(mCoordinate));
        }
    }

    @Override
    public void addToMap(MLRNMapView mapView) {
        mMapView = mapView;

        final MLRNMarkerView mlrnMarkerView = this;

        mMapView.getMapAsync(
            new OnMapReadyCallback() {
                @Override
                public void onMapReady(@NonNull MapLibreMap mapLibreMap) {
                    mMarkerViewManager = mMapView.getMarkerViewManager(mapLibreMap);

                    if (mChildView != null) {
                        mMarkerView = new MarkerView(GeoJSONUtils.toLatLng(mCoordinate), mChildView);
                        mMarkerView.setOnPositionUpdateListener(mlrnMarkerView);
                        mChildView.addOnLayoutChangeListener(mlrnMarkerView);
                        mMarkerViewManager.addMarker(mMarkerView);
                    }
                }
            }
        );
    }

    @Override
    public PointF onUpdate(PointF pointF) {
        if (mAnchor != null) {
            return new PointF(
                    pointF.x - mChildView.getWidth() * mAnchor[0],
                    pointF.y - mChildView.getHeight() * mAnchor[1]
                    );
        }
        return pointF;
    }

    @Override
    public void removeFromMap(MLRNMapView mapView) {
        if (mMarkerView != null) {
            mMarkerViewManager.removeMarker(mMarkerView);
            mChildView.removeOnLayoutChangeListener(this);
            mMarkerView.setOnPositionUpdateListener(null);
            mMarkerView = null;
            mMarkerViewManager = null;
        }
    }

    @Override
    public void onLayoutChange(View v, int left, int top, int right, int bottom, int oldLeft, int oldTop,
                               int oldRight, int oldBottom) {
        if (left != oldLeft || right != oldRight || top != oldTop || bottom != oldBottom) {
            refresh();
        }
    }
}

File: android/src/main/java/org/maplibre/reactnative/components/annotations/MLRNMarkerView.kt
package org.maplibre.reactnative.components.annotations

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.PointF
import android.view.View
import org.maplibre.geojson.Point
import org.maplibre.reactnative.components.AbstractMapFeature
import org.maplibre.reactnative.components.mapview.MLRNMapView
import org.maplibre.reactnative.utils.GeoJSONUtils

@SuppressLint("ViewConstructor")
class MLRNMarkerView(context: Context) : AbstractMapFeature(context), org.maplibre.android.plugins.markerview.MarkerView.OnPositionUpdateListener, View.OnLayoutChangeListener {
    private var mMapView: MLRNMapView? = null
    private var mChildView: View? = null
    private var mMarkerViewManager: MarkerViewManager? = null
    private var mMarkerView: MarkerView? = null
    private var mCoordinate: Point? = null
    private var mAnchor: FloatArray? = null

    override fun addView(childView: View, childPosition: Int) {
        mChildView = childView
    }

    fun setLngLat(lngLat: DoubleArray?) {
        if (lngLat == null || lngLat.size < 2) {
            return
        }
        val point = Point.fromLngLat(lngLat[0], lngLat[1])
        mCoordinate = point

        val latLng = GeoJSONUtils.toLatLng(point)
        if (latLng != null) {
            mMarkerView?.setLatLng(latLng)
        }
    }

    fun setAnchor(x: Float, y: Float) {
        mAnchor = floatArrayOf(x, y)
        this.refresh()
    }

    fun setAllowOverlap(allowOverlap: Boolean) {
        // Not implemented for Android MarkerView
    }

    fun setIsSelected(isSelected: Boolean) {
        // Not implemented for Android MarkerView
    }

    fun refresh() {
        // this will cause position to be recalculated
        val latLng = GeoJSONUtils.toLatLng(mCoordinate)
        if (latLng != null) {
            mMarkerView?.setLatLng(latLng)
        }
    }

    override fun addToMap(mapView: MLRNMapView) {
        mMapView = mapView

        val mlrnMarkerView: MLRNMarkerView = this

        mMapView?.getMapAsync { mapLibreMap ->
            val latLng = GeoJSONUtils.toLatLng(mCoordinate)
            mMarkerViewManager = mMapView?.getMarkerViewManager(mapLibreMap)
            if (latLng != null && mChildView != null) {
                mMarkerView = MarkerView(latLng, mChildView!!)
            }
            mMarkerView?.setOnPositionUpdateListener(mlrnMarkerView)
            mChildView?.addOnLayoutChangeListener(mlrnMarkerView)
            if (mMarkerView != null) {
                mMarkerViewManager?.addMarker(mMarkerView!!)
            }
        }
    }

    override fun onUpdate(pointF: PointF): PointF {
        if (mAnchor != null && mChildView != null) {
            return PointF(
                pointF.x - mChildView!!.width * mAnchor!![0],
                pointF.y - mChildView!!.height * mAnchor!![1]
            )
        }
        return pointF
    }

    override fun removeFromMap(mapView: MLRNMapView) {
        if (mMarkerView != null) {
            mMarkerViewManager?.removeMarker(mMarkerView!!)
            mChildView?.removeOnLayoutChangeListener(this)
            mMarkerView!!.setOnPositionUpdateListener(null)
            mMarkerView = null
            mMarkerViewManager = null
        }
    }

    override fun onLayoutChange(v: View, left: Int, top: Int, right: Int, bottom: Int, oldLeft: Int, oldTop: Int,
                               oldRight: Int, oldBottom: Int) {
        if (left != oldLeft || right != oldRight || top != oldTop || bottom != oldBottom) {
            refresh()
        }
    }
}
